cEP-0021.md

```diff
@@ -0,0 +1,157 @@
+then the `execute` method of Bear base class will call the
+`run_bear_from_section` method.
+
+```python
+def run_bear_from_section(self, args, kwargs):
+  ...
+  profile_bears = kwargs.get('profile_bears')
+  if profile_bears is None:
+      self.run(*args, **kwargs)
+  elif:
+      self.profile(run, *args, **kwargs)
+  ...
+```
+
+`run_bear_from_section` method checks whether bear writer invoke the profiler
+interface or not i.e., keyword arguments passed to the Bear contain the
+`profile_bears` key or not.
+
+- If *not* then `run_bear_from_section()` method will directly call
+  the `run()` method of Bear so that coala can continue its normal
+  execution.
+- Else `run_bear_from_section()` method will call `profile()` method which
+  will check the whether the value of `profile_bears` key value is
+  `True` or `Dirname`.
+
+  - If `True` then `profile()` will create a profiler file for every
+    bear of every section with a name `{section_name}{bear_name}.prof` and
+    dump it in a *current working directory.*
+
+    ```python
+    path =  os.getcwd()
+    if str(profile_bears)=='True':
+        filename = '{}_{}.prof'.format(
+            kwargs['section_name'], kwargs['bear_name'])
+        open(filename, 'w')
+        kwargs.pop('section_name')
+        kwargs.pop('bear_name')
+        cProfile.runctx('run(*args, **kwargs)',
+                        globals(), locals(), filename)
+    ```
+
+  - If `Dirname` then `profile()` method will create a profiler file for
+    every bear of every section with a name `{section_name}{bear_name}.prof` and
+    after that it check whether the specified directory exits or not.
+    If exits then it will directly dump the profiler files on the specified
+    directory. If not then it will create a subdirectory name `Dirname`
+    and dump the profiler files on that.
+
+    ```python
+    subdir = str(profile_dump)
+    filepath = os.path.join(path, subdir)
+    if os.path.isdir(filepath):
+        os.chdir(subdir)
+    else:
+        os.mkdir(os.path.join(path, subdir))
+        os.chdir(subdir)
+    filename = '{}_{}.prof'.format(
+        kwargs['section_name'], kwargs['bear_name'])
+    open(filename, 'w')
+    kwargs.pop('section_name')
+    kwargs.pop('bear_name')
+    cProfile.runctx('run(*args, **kwargs)',
+                    globals(), locals(), filename)
+    os.chdir("..")
+    ```
+
+## Profiler Output
+
+The profiler produces binary output that can be displayed and formatted by
+third party tools, which will help bear writers in creating a better performant
+bear by analysing the profiler output.
+
+For e.g.
+
+With *CLI* arguments
+
+```console
+coala --b PEPE8Bear --files <file.py> --profile-bears
+```
+
+It will create a profile file with name `cli_PEP8Bear`.
+and after that Bear writers can use different modules like
+`pstats`,`snakeviz`,`cprofilev` etc. which reads profile results from a
+file and formats them in various ways i.e.
```

Comment:
There was a way i.e., to trim the formatted data of linter but last decision was to not display or dump the formatted output directly dump the binary output and tell the user to use the extrenal tools for formatting, currently, it will use locals and globals, so it will include those

Reply:
I chaned my cEP atleast 4 times I think after a long discussion.

Reply:
> For bears which use linter executables, how will the bear profile data show time spent in the executable? it will likely be very large amounts of time, and there will be lots of stream processing occurring to pull the executable stdout/stderr back into coala.

@jayvdb Good point, we will at least see how many time is spent during `Popen` already. We can think about improving how this is displayed in the profile tools later. We need to get at least basic profiling support for all bears.

> Will it be possible to join all of the data from an entire coala run, i.e. merge the profile data for all files and all bears in all sections.?

It depends on the tool, and @RaiVaibhav told me they don't really support that in a nice fashion. Again, can be considered later.

Reply:
> There was a way i.e., to trim the formatted data of linter bears that is getting profiled but last decision was to not display or dump the formatted output directly dump the binary output and tell the user to use the extrenal tools for formatting, currently, it will use locals and globals, so it will include those

@RaiVaibhav I guess that's not what @jayvdb was aiming at^^ He wants to see some example outputs from the tools, or some general pseudo-output so we can assess what information we will get from the profiling data.